### PR TITLE
Enable PodResource API support in RTE

### PR DESCRIPTION
**Pull request by swatisehgal**
Kubelet exposes endpoint at `/var/lib/kubelet/pod-resources/kubelet.sock` for exposing information about assignment of devices to containers. It obtains this information from the internal state of the kubelet's Device Manager and returns a single PodResourcesResponse enabling monitor applications to poll for resources allocated to pods and containers on the node. This makes PodResource API a reasonable way of obtaining allocated resource information.

However, [PodResource API](https://godoc.org/k8s.io/kubernetes/pkg/kubelet/apis/podresources/v1alpha1) currently only exposes devices as the container resources (without topology info) and hence we are proposing [KEP](https://github.com/kubernetes/enhancements/pull/1884) to enhance it to expose CPU information along with device topology info.
In order to use pod-resource-api source in Resource Topology Exporter, you will need to use [this](https://github.com/kubernetes/kubernetes/pull/93243/files) patched version of kubelet implementing changes proposed in the aforementioned KEP. This will no longer be needed once the KEP and the PR are merged.

 - Added command line argument to specify source, enabling user to specify either cri or pod-resource-api
 - Created PodResourceFinder struct and supporting methods to enable support for pod-resource-api as a way
   of gathering information of allocated resources
 - Created NodeResources struct to be used for storing node resource information
 - Moved functions (updateNUMAMap(), Aggregate() and makePCI2ResourceMap()) used by both crifinder.go and
   podresourcefinder.go to finder.go
 - Narrowed down volume mounts to the required subtree
 - Updated README

### CODE REVIEW

**Comment by swatisehgal** [comment]
This PR has been tested and is working as expected.

**Comment by swatisehgal** [comment]
@fromanirh PTAL, Kindly review this PR and let me know your feedback, Thanks.

File: go.mod [hunk 1]
```diff
@@ -47,7 +47,7 @@ replace (
 	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.18.6
 	k8s.io/kubectl => k8s.io/kubectl v0.18.6
 	k8s.io/kubelet => k8s.io/kubelet v0.18.6
-	k8s.io/kubernetes => k8s.io/kubernetes v1.18.6
+	k8s.io/kubernetes => ../../../k8s.io/kubernetes
```
Comment (swatisehgal) [inline go.mod:50]: This points to a local kubernetes repository (with changes implemented [here](https://github.com/kubernetes/kubernetes/pull/93243#pullrequestreview-464074662))

File: main.go [hunk 2]
```diff
@@ -44,24 +43,33 @@ func main() {
 	}
 
 	// Get new finder instance
-	instance, err := finder.NewFinder(args, pci2ResMap)
-	if err != nil {
-		log.Fatalf("Failed to initialize Finder instance: %v", err)
+	var finderInstance finder.Finder
+	if args.Source == "cri"{
```
Comment (ffromani) [inline main.go]: let's use constants for source types

File: main.go [hunk 3]
```diff
@@ -44,24 +43,33 @@ func main() {
...
+		finderInstance, err = finder.NewCRIFinder(args, pci2ResMap)
+		if err != nil {
+			log.Fatalf("Failed to initialize Finder instance: %v", err)
+		}
+	}else if args.Source == "pod-resource-api"{
+		finderInstance, err = finder.NewPodResourceFinder(args, pci2ResMap)
+		if err != nil {
+			log.Fatalf("Failed to initialize Finder instance: %v", err)
+		}
```
Comment (ffromani) [inline main.go]: no need to duplicate this check: just move it outside the if/else

File: main.go [hunk 4]
```diff
@@ -76,29 +84,35 @@ func main() {
...
 	}
 	usage := fmt.Sprintf(`Usage:
-  %s [--sleep-interval=<seconds>] [--cri-path=<path>] [--watch-namespace=<namespace>] [--sysfs=<mountpoint>] [--sriov-config-file=<path>] [--container-runtime=<runtime>] [--kubelet-config-file=<path>]
+  %s [--sleep-interval=<seconds>] [--source=<path>] [--container-runtime=<runtime>] [--cri-socket=<path>] [--podresources-socket=<path>] [--watch-namespace=<namespace>] [--sysfs=<mountpoint>] [--sriov-config-file=<path>] [--kubelet-config-file=<path>]
   %s -h | --help
   Options:
   -h --help                       Show this screen.
-  --container-runtime=<runtime>   Container Runtime to be used (containerd|cri-o). [Default: %v]
-  --cri-path=<path>               CRI Endpoint file path to use. [Default: %v]
+	--source=<source>								Evaluation source to be used (pod-resource-api|cri). [Default: %v]
```
Comment (ffromani) [inline main.go:104]: inconsistent spacing (also on line 104 below)

File: manifests/resource-topology-exporter-ds.yaml [hunk 5]
```diff
@@ -69,6 +69,8 @@ spec:
             mountPath: "/host-etc"
           - mountPath: /etc/sriov-config
             name: my-sriov-config-vol
+          - name: host-var
+            mountPath: "/host-var"
```
Comment (ffromani) [inline manifests/resource-topology-exporter-ds.yaml]: let's narrow down as much as possible the subtree we mount, especially if this needs to be mounted in RW

File: pkg/finder/podresource.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+// podResourcesServer implements PodResourcesListerServer
+type podResourcesServer struct {
+	podsProvider    PodsProvider
+	devicesProvider DevicesProvider
+}
+
+// NewPodResourcesServer returns a PodResourcesListerServer which lists pods provided by the PodsProvider
+// with device information provided by the DevicesProvider
+func NewPodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider) podresourcesapi.PodResourcesListerServer {
```
Comment (ffromani) [inline pkg/finder/podresource.go]: could you please elaborate about why we need the server side here? Isn't RTE the consumer here?

File: pkg/finder/podresourcefinder.go [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+// 	Snapshotter    string              `json:"snapshotter"`
+// 	RuntimeType    string              `json:"runtimeType"`
+// 	RuntimeOptions interface{}         `json:"runtimeOptions"`
+// 	Config         *pb.ContainerConfig `json:"config"`
+// 	RuntimeSpec    *runtimespec.Spec   `json:"runtimeSpec"`
+// }
+
+
+func NewPodResourceFinder(args Args, pciResMapConf map[string]string) (Finder, error) {
+	finderInstance := &PodResourceFinder{
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go:29]: if we need all this (up until line 114), can we factor this out and avoid duplication?

File: pkg/finder/podresourcefinder.go [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+	}
+	for _, pciAddr := range ri.Data {
+		nodeNum, ok := cpf.pciAddr2NUMAID[pciAddr]
+		if !ok {
+			log.Printf("unknown PCI address: %q", pciAddr)
+			continue
+		}
+		numaData[nodeNum][ri.Name]--
+	}
+}
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: this is the same code we have in the CRI counterpart, right? can we factor it out somehow?

File: pkg/finder/podresourcefinder.go [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+		for _, container := range podResource.GetContainers() {
+		contRes := ContainerResources{
+			Name: container.Name,
+		}
+		devs := make(map[string][]string)
+		for _, device := range container.GetDevices() {
+			devs[device.ResourceName] = device.DeviceIds
+		}
+		// cpuList := container.GetCPUs()
+		cpuList := []int32{1,2,3,4,5,6,7,8,9,10}
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: hardcoded values? Is this temporary?

File: pkg/finder/podresourcefinder.go [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+			NUMAID:    nodeNum,
+			Resources: make(v1.ResourceList),
+		}
+		for name, intQty := range resList {
+			numaRes.Resources[name] = *resource.NewQuantity(intQty, resource.DecimalSI)
+		}
+		perNumaRes = append(perNumaRes, numaRes)
+	}
+	return perNumaRes
+}
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: can we factor this out somehow?

File: pkg/finder/podresourcefinder.go [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+		}
+
+
+	}
+		resInfos = append(resInfos, ResourceInfo{
+			Name: v1.ResourceName(resName),
+			Data: pciAddrs,
+		})
+	return resInfos
+}
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: same

**Review by ffromani — COMMENTED** [review]
ok, I see some of the comments I have are already addressed in the second patch. Feel free to squash it in just one patch

File: main.go (hunk 2 above)
Comment (swatisehgal) [reply main.go]: Done

File: main.go (hunk 3 above)
Comment (swatisehgal) [reply main.go]: Done

File: pkg/finder/podresource.go (hunk 6 above)
Comment (swatisehgal) [reply pkg/finder/podresource.go]: This file was removed as the implementation progressed, the commits have been squashed now

File: pkg/finder/podresourcefinder.go (hunk 7 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go:29]: +1. I have moved this to finder.go. Let me know if looks better now

File: pkg/finder/podresourcefinder.go (hunk 8 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: Same as above

File: pkg/finder/podresourcefinder.go (hunk 9 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: yes, was temporary when I wasn't using the patched version of kubelet, later moved to obtaining the cpus allocated from getCPU() function

File: pkg/finder/podresourcefinder.go (hunk 10 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: Ack

File: pkg/finder/podresourcefinder.go (hunk 11 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: Ack

File: main.go (hunk 4 above)
Comment (swatisehgal) [reply main.go:104]: addressed now

File: main.go [hunk 12]
```diff
@@ -76,29 +85,29 @@ func main() {
...
+  --container-runtime=<runtime>   Container Runtime to be used (containerd|cri-o).
+  --cri-socket=<path>             CRI Socket path to use.
+	--podresources-socket=<path>    Pod Resource Socket path to use.
```
Comment (ffromani) [inline main.go:107]: ...and here.

File: main.go [hunk 13]
```diff
@@ -118,12 +127,40 @@ func argsParse(argv []string) (finder.Args, error) {
...
+			args.CRISocketPath = path
+		}
+		if _, ok := arguments["--podresources-socket"].(string); ok {
+			return args, fmt.Errorf("No need to specify Pod Resource socket path in case CRI is specified as the source")
+		}
+		runtime := arguments["--container-runtime"].(string)
+		if !(runtime == ContainerdRuntime || runtime == CRIORuntime) {
+			return args, fmt.Errorf("invalid --container-runtime specified")
+		}
+		args.ContainerRuntime = runtime
```
Comment (ffromani) [inline main.go:164]: a little hard to follow but we can clean it up later

File: manifests/resource-topology-exporter-ds.yaml [hunk 14]
```diff
@@ -53,22 +53,22 @@ spec:
...
         image: quay.io/swsehgal/resource-topology-exporter:latest
-#        command:
-#        - /usr/local/bin/resource-topology-exporter
-#        - --sleep-interval=1s
         command:
-        - sleep
-        args:
-        - "1000000"
+        - /usr/local/bin/resource-topology-exporter
+        - --watch-namespace=rte
```
Comment (ffromani) [inline manifests/resource-topology-exporter-ds.yaml:58]: in a future PR, I believe we should start watching all the namespaces

File: pkg/finder/podresourcefinder.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+	"k8s.io/api/core/v1"
+
+	podresources "k8s.io/kubernetes/pkg/kubelet/apis/podresources"
+	podresourcesapi "k8s.io/kubernetes/pkg/kubelet/apis/podresources/v1alpha1"
+)
+
+const (
+	defaultPodResourcesTimeout = 10 * time.Second
+	defaultPodResourcesMaxSize = 1024 * 1024 * 16 // 16 Mb
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go:19]: Please document how we decided these values (is fine if this is an initial estimation, but just document that)

File: pkg/finder/podresourcefinder.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+)
+
+type PodResourceFinder struct {
+	args              Args
+	podResourceClient podresourcesapi.PodResourcesListerClient
+	NodeResourceData  *NodeResources
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: why public? (capital letter identifier - can be fixed later)

File: pkg/finder/crifinder.go [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+	CRIORuntime        = "cri-o"
+)
+
+type CGroupPathTranslator func(sysfs, cgroupPath string) string
+type criFinder struct {
+	args             Args
+	conn             *grpc.ClientConn
+	client           pb.RuntimeServiceClient
+	cgroupTranslate  CGroupPathTranslator
+	NodeResourceData *NodeResources
```
Comment (ffromani) [inline pkg/finder/crifinder.go]: why public? (capital letter identifier - can be fixed later)

File: pkg/finder/crifinder.go [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+	var err error
+
+	// At this stage, we only support containerd and cri-o
+	if args.ContainerRuntime == ContainerdRuntime {
+		finderInstance.cgroupTranslate = containerDCGroupPathTranslate
+	} else {
+		//cri-o
+		finderInstance.cgroupTranslate = crioCGroupPathTranslate
+	}
+	finderInstance.NodeResourceData, err = NewNodeResources(finderInstance.args.SysfsRoot, pciResMapConf)
```
Comment (ffromani) [inline pkg/finder/crifinder.go]: `NodeResources` code seems to be lost in rebase - if not could you please point me to it?

File: pkg/finder/crifinder.go [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+
+	return finderInstance, nil
+}
+
+func getAddressAndDialer(endpoint string) (string, func(ctx context.Context, addr string) (net.Conn, error), error) {
+	return util.GetAddressAndDialer(endpoint)
+}
+
+func (f *criFinder) GetNodeResourceData() *NodeResources {
+	return f.NodeResourceData
```
Comment (ffromani) [inline pkg/finder/crifinder.go]: I think `Scan()` should return this data; IOW I don't see the value of this extra getter. Can be done later.

File: pkg/finder/podresourcefinder.go [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+			contRes := ContainerResources{
+				Name: container.Name,
+			}
+			devs := make(map[string][]string)
+			for _, device := range container.GetDevices() {
+				devs[device.ResourceName] = device.DeviceIds
+			}
+			cpuList := container.GetCpuIds()
+			contRes.Resources = append(contRes.Resources, makeCPUResourceInfo(cpuList)...)
+			contRes.Resources = append(contRes.Resources, makePCIDeviceResourceInfo(devs, f.NodeResourceData.pci2ResourceMap)...)
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: so are deviceIds guaranteed to be PCI addresses? Always, no exceptions? This is a key requirement that needs to be documented (oneliner is enough, but we need it)

File: pkg/finder/podresourcefinder.go [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+	return podResData, nil
+}
+
+func makeCPUResourceInfo(cpus []uint32) []ResourceInfo {
+	var ret []string
+	for _, cpuID := range cpus {
+		ret = append(ret, fmt.Sprintf("%d", cpuID))
+	}
+
+	fmt.Printf("makeCPUResourceInfo: data %v", ret)
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: please remove

File: pkg/finder/podresourcefinder.go [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+			resName, ok = pci2ResMap[devId]
+			if !ok {
+				continue
+			}
+			pciAddrs = append(pciAddrs, devId)
+		}
+
+	}
+
+	fmt.Printf("makePCIDeviceResourceInfo: data %v", pciAddrs)
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go]: same, please remove

File: pkg/finder/podresourcefinder.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+	resInfos = append(resInfos, ResourceInfo{
+		Name: v1.ResourceName(resName),
+		Data: pciAddrs,
+	})
+	return resInfos
+}
```
Comment (ffromani) [inline pkg/finder/podresourcefinder.go:135]: it seems to me that we have duplicated code. Can be fixed later.

**Review by ffromani — COMMENTED** [review]
thanks, getting better, still few questions inline.

File: manifests/resource-topology-exporter-ds.yaml (hunk 14 above)
Comment (swatisehgal) [reply manifests/resource-topology-exporter-ds.yaml:58]: By default all the namespaces are watched, for demo and testing purpose I check for pods in a certain namespace (rte in this case). I can remove --watch-namespace from the yaml if you think that is more appropriate?

File: pkg/finder/podresourcefinder.go (hunk 15 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go:19]: Obtained them from [here](https://github.com/kubernetes/kubernetes/blob/82baa26905c94398a0d19e1b1ecf54eb8acb6029/test/e2e_node/util.go#L70). Can add a comment in the code referring to this link

File: pkg/finder/podresourcefinder.go (hunk 16 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: updating it to private, I am not sure why I made it public in the first place :)

File: pkg/finder/crifinder.go (hunk 19 above)
Comment (swatisehgal) [reply pkg/finder/crifinder.go]: Scan returns PodResourceData which is information about all the allocated resources to the pods whereas this getter returns nodeResourceData (pointer to instance NodeResources):
```
type NodeResources struct {
	pciDevs         *pcidev.PCIDevices
	cpus            *cpus.CPUs
	cpuID2NUMAID    map[int]int
	pciAddr2NUMAID  map[string]int
	perNUMACapacity map[int]map[v1.ResourceName]int64
	// pciaddr -> resourcename
	pci2ResourceMap map[string]string
}
```

nodeResourceData stores information of the resources available (CPUs obtained after scanning sysfs and devices obtained by using sr-iov config file) and  on the node on which the daemon is running. But, since we are assuming that the resources on the nodes will remain the same during the process life cycle, we can do this get this information once in main rather than having it in both criFinder and podResourceFinder structs. So, moving code pertaining to to NodeResources in a separate file and instantiating NodeResourceData in main now.

File: pkg/finder/podresourcefinder.go (hunk 20 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go]: There was a comment [here](https://github.com/swatisehgal/resource-topology-exporter/blob/pod-resource-api/pkg/finder/podresourcefinder.go#L126) briefly touching on this but will update to capture this in better words

File: pkg/finder/podresourcefinder.go (hunk 23 above)
Comment (swatisehgal) [reply pkg/finder/podresourcefinder.go:135]: So, the code seems duplicate but if you look carefully there are differences because of difference in the format in which data is obtained from CRI and podresource API. 

- For cpu information, CRI returns CPUSet whereas PodResource api returns []unit32

- For device information, in CRI we have a separate comma separated device PCI addresses whereas in case of pod resource api we get a slice of devices.

But, I certainly see too this can be improved by finding least common denominator and/or a bit of type conversion. Will address this in a separate PR later

File: main.go (hunk 13 above)
Comment (swatisehgal) [reply main.go:164]: Added a few comments to explain the code, hope it helps.

File: main.go [hunk 24]
```diff
@@ -63,14 +63,19 @@ func main() {
 		log.Fatalf("Failed to initialize crdExporter instance: %v", err)
 	}
 
+	// CAUTION: these resources are expected to change rarely - if ever. So we are intentionally do this once during the process lifecycle.
+	nodeResourceData, err := finder.NewNodeResources(args.SysfsRoot, pci2ResMap)
+	if err != nil {
+		log.Fatalf("Failed to obtain node resource information: %v", err)
+	}
```
Comment (ffromani) [inline main.go:70]: nice!

File: pkg/finder/podresourcefinder.go (hunk 20 above)
Comment (ffromani) [reply pkg/finder/podresourcefinder.go]: Yes, but from CRI API we can have reasonnable (even though not actually trusted) expectations, because we know how device plugin work; podresources API can have a weaker guarantee however.
Since one of the driving factors of this PR is moving past the PoC stage, we need to have those guarantees in place, or at least we need to go in that direction

**Comment by ffromani** [comment]
It's a bit awkward because the lack of tests and, most important, the fact we depend on a patched kube tree, make hard to actually try out and have good confidence in this PR. Everything considered I'm inclined to move forward and iterate later.
Having good tests is more and more a priority.
/lgtm

**Comment by swatisehgal** [comment]
I see the concerns about patched kube tree, but I have tried to capture that in the README in case someone wants to try it out. Also, since we are supporting CRI as a source it could be a way of getting RTE running (without needing to patch kubelet). Re: Testing, I see the concerns and will work to improve the overall testing of this project in the next few weeks.

Thanks for great feedback on this PR!
